suppression-list/csv2optout.py

```diff
@@ -0,0 +1,208 @@
+DROP optouts_{tmp_suffix} CASCADE;
+"""
+
+SQL_BATCH = """
+BEGIN;
+
+CALL raise_notice('Batch {batch}/{batch_count}');
+
+UPDATE emails
+  SET update_timestamp = tmp.ts,
+      has_opted_out_of_email = true,
+      unsubscribe_reason = tmp.unsubscribe_reason
+  FROM optouts_{tmp_suffix} tmp
+  WHERE tmp.email_id = emails.email_id
+    AND tmp.idx > {start_idx} AND tmp.idx <= {end_idx}
+    -- Do not overwrite reason if user opted-out in the mean time
+    AND has_opted_out_of_email IS NOT true;
+
+INSERT INTO pending_acoustic(email_id, retry)
+  SELECT email_id, 0 FROM optouts_{tmp_suffix}
+    WHERE {schedule_sync} AND idx > {start_idx} AND idx <= {end_idx};
+
+DELETE FROM optouts_{tmp_suffix}
+  WHERE idx > {start_idx} AND idx <= {end_idx};
+
+COMMIT;
+
+SELECT pg_sleep({sleep_seconds});
+"""
+
+
+def chunks(l, n):
+    for i in range(0, len(l), n):
+        yield l[i : i + n]
+
+
+def writefile(path, content):
+    with open(path, "w") as f:
+        f.write(content)
+    logger.info(f"{path!r} written.")
+
+
+@click.command()
+@click.argument("csv_path", type=click.Path(exists=True))
+@click.option(
+    "--check-input-rows", default=1000, help="Number of rows to check from input CSV."
+)
+@click.option("--batch-size", default=10000, help="Number of updates per commit.")
+@click.option("--files-count", default=5, help="Number of SQL files")
+@click.option("--sleep-seconds", default=0.1, help="Wait between batches")
+@click.option(
+    "--schedule-sync", default=False, help="Mark update emails as pending sync"
+)
+def main(
+    csv_path, check_input_rows, batch_size, files_count, sleep_seconds, schedule_sync
+) -> int:
+    #
+    # Inspect CSV input.
+    #
+    with open(csv_path) as f:
+        csv_rows_count = sum(1 for _ in f)
+        f.seek(0)
+        delimiter = csv.Sniffer().sniff(f.read(1024)).delimiter
+        f.seek(0)
+        has_headers = csv.Sniffer().has_header(f.read(1024))
+        f.seek(0)
+
+        # Check format of X entries.
+        reader = csv.reader(f)
+        if has_headers:
+            next(reader)
+        for i, row in enumerate(reader):
+            if i >= check_input_rows:
+                break
+            try:
+                email, date, reason = row
+                assert "@" in email
+                assert re.match(r"\d{4}-\d{2}-\d{2} \d{2}:\d{2} (AM|PM)", date)
+            except (AssertionError, ValueError):
+                raise ValueError(f"Line '{row}' does not look right")
+
+    batch_count = 1 + csv_rows_count // batch_size
+    logger.info(
+        f"{csv_rows_count} entries, {batch_count} batches of {batch_size} updates per commit."
+    )
+
+    #
+    # Prepare SQL files
+    #
+    now = datetime.now(tz=timezone.utc)
+    tmp_suffix = now.strftime("%Y%m%dT%H%M")
+    batch_commands = []
+    for i in range(batch_count):
+        start_idx = i * batch_size
+        end_idx = (i + 1) * batch_size
+        batch_commands.append(
+            SQL_BATCH.format(
```

Comment:
Possible injection flaw. This is a one-off and we aren't going to reuse it in application code, so not gonna push hard. But ideally this would be a parameterized query or database function.

Reply:
Yeah, well, the script outputs a sql file that is meant to be executed in `psql`